data_safe_haven/infrastructure/programs/sre/firewall.py

```diff
@@ -0,0 +1,270 @@
+                                    protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTPS,
+                                )
+                            ],
+                            source_addresses=props.subnet_user_services_software_repositories_prefixes,
+                            target_fqdns=PermittedDomains.SOFTWARE_REPOSITORIES_PYTHON,
+                        ),
+                    ],
+                ),
+                network.AzureFirewallApplicationRuleCollectionArgs(
+                    action=network.AzureFirewallRCActionArgs(
+                        type=network.AzureFirewallRCActionType.ALLOW
+                    ),
+                    name="workspaces",
+                    priority=FirewallPriorities.SRE_WORKSPACES,
+                    rules=[
+                        network.AzureFirewallApplicationRuleArgs(
+                            description="Allow external Ubuntu keyserver requests",
+                            name="AllowUbuntuKeyserver",
+                            protocols=[
+                                network.AzureFirewallApplicationRuleProtocolArgs(
+                                    port=int(Ports.CLAMAV),
+                                    protocol_type=network.AzureFirewallApplicationRuleProtocolType.HTTP,
+                                ),
+                            ],
+                            source_addresses=props.subnet_workspaces_prefixes,
+                            target_fqdns=PermittedDomains.UBUNTU_KEYSERVER,
+                        ),
+                    ],
+                ),
+            ],
+            azure_firewall_name=f"{stack_name}-firewall",
+            ip_configurations=[
+                network.AzureFirewallIPConfigurationArgs(
+                    name="FirewallIpConfiguration",
+                    public_ip_address=network.SubResourceArgs(id=public_ip.id),
+                    subnet=network.SubResourceArgs(id=props.subnet_firewall_id),
+                )
+            ],
+            location=props.location,
+            management_ip_configuration=network.AzureFirewallIPConfigurationArgs(
+                name="FirewallManagementIpConfiguration",
+                public_ip_address=network.SubResourceArgs(id=public_ip_management.id),
+                subnet=network.SubResourceArgs(id=props.subnet_firewall_management_id),
+            ),
+            resource_group_name=props.resource_group_name,
+            sku=network.AzureFirewallSkuArgs(
+                name=network.AzureFirewallSkuName.AZF_W_V_NET,
+                tier=network.AzureFirewallSkuTier.BASIC,
+            ),
+            opts=child_opts,
+            tags=child_tags,
+        )
+
+        # Retrieve the private IP address for the firewall
+        private_ip_address = firewall.ip_configurations.apply(
+            lambda cfgs: (
+                "" if not cfgs else next(cfg.private_ip_address for cfg in cfgs)
+            )
+        )
```

Comment:
This just means `private_ip_address` is a generator rather than a list? Or it just gives the first item?

Reply:
- `private_ip_address` is an `Output[str]`
- `cfgs` is a `Sequence[AzureFirewallIPConfigurationResponse] | None`
- `cfg.private_ip_address` is a `str`

so inside the `apply` loop we assign `""` if `cfgs` is `None` or the first `private_ip_address` among the `cfgs` if they exist (there should only be one)

Reply:
Is it better to use `next` rather than `cfgs[0].private_ip_address`?